dsl/pace/dsl/dace/utils.py

```diff
@@ -120,4 +128,119 @@ def evaluate(expr):
+            allocations[arr.storage].details.append(
+                ArrayReport(
+                    name=aname,
+                    total_size_in_bytes=array_size_in_bytes,
+                    referenced=ref,
+                    transient=arr.transient,
+                    pool=arr.pool,
+                    top_level=False,
+                )
+            )
+            if ref:
+                allocations[arr.storage].referenced_in_bytes += array_size_in_bytes
+            else:
+                allocations[arr.storage].unreferenced_in_bytes += array_size_in_bytes
+
+        elif sd is sdfg:
+            if arr.pool:
+                allocations[arr.storage].in_pooled_in_bytes += array_size_in_bytes
+            allocations[arr.storage].details.append(
+                ArrayReport(
+                    name=aname,
+                    total_size_in_bytes=array_size_in_bytes,
+                    referenced=ref,
+                    transient=arr.transient,
+                    pool=arr.pool,
+                    top_level=True,
+                )
+            )
+            allocations[arr.storage].top_level_in_bytes += array_size_in_bytes
+            if ref:
+                allocations[arr.storage].referenced_in_bytes += array_size_in_bytes
+            else:
+                allocations[arr.storage].unreferenced_in_bytes += array_size_in_bytes
+
+    report = f"{sdfg.name}:\n"
+    for storage, allocs in allocations.items():
+        alloc_in_mb = float(allocs.referenced_in_bytes / (1024 * 1024))
+        unref_alloc_in_mb = float(allocs.unreferenced_in_bytes / (1024 * 1024))
+        in_pooled_in_mb = float(allocs.in_pooled_in_bytes / (1024 * 1024))
+        toplvlalloc_in_mb = float(allocs.top_level_in_bytes / (1024 * 1024))
+        if alloc_in_mb or toplvlalloc_in_mb > 0:
+            report += (
+                f"{storage}:\n"
+                f"  Alloc ref {alloc_in_mb:.2f} mb\n"
+                f"  Alloc unref {unref_alloc_in_mb:.2f} mb\n"
+                f"  Pooled {in_pooled_in_mb:.2f} mb\n"
+                f"  Top lvl alloc: {toplvlalloc_in_mb:.2f}mb\n"
+            )
+            if detail_report:
+                report += "\n"
+                report += "  Referenced\tTransient   \tPooled\tTotal size(mb)\tName\n"
+                for detail in allocs.details:
+                    size_in_mb = float(detail.total_size_in_bytes / (1024 * 1024))
+                    ref_str = "     X     " if detail.referenced else "           "
+                    transient_str = "     X     " if detail.transient else "           "
+                    pooled_str = "     X     " if detail.pool else "           "
+                    report += (
+                        f" {ref_str}\t{transient_str}"
+                        f"\t   {pooled_str}"
+                        f"\t   {size_in_mb:.2f}"
+                        f"\t   {detail.name}\n"
+                    )
+
+    return report
```

Comment:
Just a suggestion: it would be more flexible to separate the counting memory from the report generation, i.e. split this into two functions. The first would create something like `[(name, size)]` for an SDFG, then the next one would the english representation.

Reply:
I am prepping another PR that renames that tool and add a kernel timing analysis. I'll break it up there